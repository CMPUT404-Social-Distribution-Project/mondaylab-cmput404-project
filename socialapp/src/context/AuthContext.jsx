import React, { createContext, useState, useEffect } from "react";
import jwt_decode from "jwt-decode";
import { useNavigate } from "react-router-dom";

const AuthContext = createContext();

export default AuthContext;

export const AuthProvider = ({ children }) => {
  const [authTokens, setAuthTokens] = useState(() =>
    localStorage.getItem("authTokens")
      ? JSON.parse(localStorage.getItem("authTokens"))
      : null
  );
  const [user, setUser] = useState(() =>
    localStorage.getItem("authTokens")
      ? jwt_decode(localStorage.getItem("authTokens"))
      : null
  );
  const [loading, setLoading] = useState(true);
  const [loginLoading, setLoginLoading] = useState(null);
  const baseURL = "https://cs404-project.herokuapp.com/service";

  const navigate = useNavigate();

  const loginUser = async (displayName, password) => {
    setLoginLoading(true);
    const response = await fetch(baseURL + "/auth/login/", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        displayName,
        password,
      }),
    });
    await response.json().then(function (data) {
      if (response.status === 200) {
        setAuthTokens(data);
        setUser(jwt_decode(data.access));
        localStorage.setItem("authTokens", JSON.stringify(data));
        localStorage.setItem("loggedInUser", JSON.stringify(data.user));
        setLoginLoading(false);
      } else {
        alert("ERROR: " + data);
        setLoginLoading(null);
      }
    });
  };

  const registerUser = async (displayName, password, github) => {
    const response = await fetch(baseURL + "/auth/register/", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        displayName,
        password,
        github,
      }),
    });

    if (response.status === 201) {
      navigate("/login");
    } else {
      response.json().then(function (value) {
        alert("ERROR: " + value);
      });
    }
  };

  const logoutUser = () => {
    setAuthTokens(null);
    setUser(null);
    localStorage.removeItem("authTokens");
    localStorage.removeItem("user_id");
    navigate("/login");
  };

  const contextData = {
    user,
    setUser,
    authTokens,
    setAuthTokens,
    registerUser,
    loginUser,
    logoutUser,
    baseURL,
  };

  useEffect(() => {
    if (authTokens) {
      setUser(jwt_decode(authTokens.access));
      localStorage.setItem("user_id", user.user_id.split("/").pop());
    }
    setLoading(false);
  }, [authTokens, loading]);

  // Waits user to be set before sending user to main page. Otherwise user would be null
  // and the homepage will show nothing because it would fetch to /service/authors/null/posts/.
  // This is due to async between setting the user in loginUser & the above useEffect hook
  useEffect(() => {
    if (loginLoading !== null && loginLoading === false) {
      navigate("/stream");
    }
    setLoginLoading(null);
  }, [loginLoading]);

  return (
    <AuthContext.Provider value={contextData}>
      {loading ? null : children}
    </AuthContext.Provider>
  );
};
