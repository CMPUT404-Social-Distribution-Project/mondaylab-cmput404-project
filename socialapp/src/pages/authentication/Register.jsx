import React, { useContext } from 'react';
import "./Auth.css";
import { Formik, Form } from 'formik';
import TextField from '../../components/TextField';
import * as Yup from 'yup';
import Logo from "../../des/logos/logo.svg";
import LogoText from "../../des/logos/logo-text.svg";
import AuthContext from "../../context/AuthContext";
import { useNavigate } from "react-router-dom";


export default function Register() {
    const validate = Yup.object().shape({
        username: Yup.string().required('Required').max(26, 'Invalid username'),
        password: Yup.string().required(),
        });

    const { registerUser } = useContext(AuthContext);
    
    const navigate = useNavigate();

  return (
    <div className="Auth-form-container">
    <div className="introduction">
        <div className="Logo-container">
            <img src={Logo} />
            <img 
            id="logo-text"
            src={LogoText} 
            />
        </div>
        <div id="intro-text" className="text-left">
            Welcome to SocialDistribution,<br/> your one stop shop for everything social media.
        </div>
    </div>
            <Formik
            initialValues={{
                username: '',
                password: '',
                github: '',
            }}
            validationSchema={validate}
            onSubmit={(values)=> {
                console.log(values);
                const username = values.username;
                const password = values.password;
                const github = values.github;
                registerUser(username, password);
            }}
        >
            {formik => (
                <div>
                <Form className='form'>
                    <h3 className="Auth-form-title">Signup</h3>
                    <div className="text-center">
                        Already registered?{" "}
                        <span id="reg-login" class="link-primary" onClick={() => navigate("/login")}>
                            Login
                        </span>
                    </div>
                    <TextField label="Username" name="username" type="text"/>
                    <TextField label="Password" name="password" type="password"/>
                    <TextField label="Confirm Password" name="confirmPass" type="password"/>
                    <TextField label="GitHub" name="github" type="text"/>
                    <button id="submit-button" type="submit" className="btn btn-primary">Submit</button>
                </Form>
            </div>
            )}
        </Formik>
</div>
  )
}
