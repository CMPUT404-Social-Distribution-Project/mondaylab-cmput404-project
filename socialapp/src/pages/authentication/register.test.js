
import React from "react";
import renderer from "react-test-renderer";
import { fireEvent, render } from '@testing-library/react';
import Register from './Register';
import { AuthProvider } from "../../context/AuthContext";
import { Route, BrowserRouter } from "react-router-dom";

describe('Testing Register:', () => {
    it("renders correctly", () => {
        const tree = render(<BrowserRouter><Register /></BrowserRouter>);
        expect(tree).toMatchSnapshot();
    });
});