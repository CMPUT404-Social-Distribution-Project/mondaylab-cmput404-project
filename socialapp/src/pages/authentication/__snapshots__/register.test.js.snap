// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Register: renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Auth-form-container"
      >
        <div
          class="introduction"
        >
          <div
            class="Logo-container"
          >
            <img
              alt="logo"
              src="logo.svg"
            />
            <img
              alt="logo-text"
              id="logo-text"
              src="logo-text.svg"
            />
          </div>
          <div
            class="text-left"
            id="intro-text"
          >
            Welcome to SocialDistribution,
            <br />
             your one stop shop for everything social media.
          </div>
        </div>
        <div>
          <form
            action="#"
            class="form"
          >
            <h3
              class="Auth-form-title"
            >
              Signup
            </h3>
            <div
              class="text-center"
            >
              Already registered?
              <span
                class="link-primary"
                id="reg-login"
              >
                Login
              </span>
            </div>
            <div
              class="mb-2 mt-4"
            >
              <label
                for="username"
              >
                Username
              </label>
              <input
                autocomplete="off"
                class="form-control shadow-none false"
                name="username"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-2 mt-4"
            >
              <label
                for="password"
              >
                Password
              </label>
              <input
                autocomplete="off"
                class="form-control shadow-none false"
                name="password"
                type="password"
                value=""
              />
            </div>
            <div
              class="mb-2 mt-4"
            >
              <label
                for="confirmPass"
              >
                Confirm Password
              </label>
              <input
                autocomplete="off"
                class="form-control shadow-none false"
                name="confirmPass"
                type="password"
                value=""
              />
            </div>
            <div
              class="mb-2 mt-4"
            >
              <label
                for="github"
              >
                GitHub
              </label>
              <input
                autocomplete="off"
                class="form-control shadow-none false"
                name="github"
                type="text"
                value=""
              />
            </div>
            <button
              class="btn btn-primary"
              id="submit-button"
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Auth-form-container"
    >
      <div
        class="introduction"
      >
        <div
          class="Logo-container"
        >
          <img
            alt="logo"
            src="logo.svg"
          />
          <img
            alt="logo-text"
            id="logo-text"
            src="logo-text.svg"
          />
        </div>
        <div
          class="text-left"
          id="intro-text"
        >
          Welcome to SocialDistribution,
          <br />
           your one stop shop for everything social media.
        </div>
      </div>
      <div>
        <form
          action="#"
          class="form"
        >
          <h3
            class="Auth-form-title"
          >
            Signup
          </h3>
          <div
            class="text-center"
          >
            Already registered?
            <span
              class="link-primary"
              id="reg-login"
            >
              Login
            </span>
          </div>
          <div
            class="mb-2 mt-4"
          >
            <label
              for="username"
            >
              Username
            </label>
            <input
              autocomplete="off"
              class="form-control shadow-none false"
              name="username"
              type="text"
              value=""
            />
          </div>
          <div
            class="mb-2 mt-4"
          >
            <label
              for="password"
            >
              Password
            </label>
            <input
              autocomplete="off"
              class="form-control shadow-none false"
              name="password"
              type="password"
              value=""
            />
          </div>
          <div
            class="mb-2 mt-4"
          >
            <label
              for="confirmPass"
            >
              Confirm Password
            </label>
            <input
              autocomplete="off"
              class="form-control shadow-none false"
              name="confirmPass"
              type="password"
              value=""
            />
          </div>
          <div
            class="mb-2 mt-4"
          >
            <label
              for="github"
            >
              GitHub
            </label>
            <input
              autocomplete="off"
              class="form-control shadow-none false"
              name="github"
              type="text"
              value=""
            />
          </div>
          <button
            class="btn btn-primary"
            id="submit-button"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
